provider/buffered/provider.go

```diff
@@ -0,0 +1,244 @@
+	stopProvidingOp
+	// lastOp is used for array sizing.
+	lastOp
+)
+
+var _ internal.Provider = (*SweepingProvider)(nil)
+
+// SweepingProvider implements a buffered provider that queues operations and
+// processes them asynchronously in batches.
+type SweepingProvider struct {
+	closeOnce sync.Once
+	done      chan struct{}
+	closed    chan struct{}
+	provider  internal.Provider
+	queue     *dsqueue.DSQueue
+	batchSize int
+}
+
+// New creates a new SweepingProvider that wraps the given provider with
+// buffering capabilities. Operations are queued and processed asynchronously
+// in batches for improved performance.
+func New(prov internal.Provider, ds datastore.Batching, opts ...Option) *SweepingProvider {
+	cfg := getOpts(opts)
+	s := &SweepingProvider{
+		done:   make(chan struct{}),
+		closed: make(chan struct{}),
+
+		provider: prov,
+		queue: dsqueue.New(ds, cfg.dsName,
+			dsqueue.WithDedupCacheSize(0), // disable deduplication
+			dsqueue.WithIdleWriteTime(cfg.idleWriteTime),
+		),
+		batchSize: cfg.batchSize,
+	}
+	go s.worker()
+	return s
+}
+
+// Close stops the provider and releases all resources.
+//
+// It waits for the worker goroutine to finish processing current operations
+// and closes the underneath provider. The queue current state is persisted on
+// the datastore.
+func (s *SweepingProvider) Close() error {
+	var err error
+	s.closeOnce.Do(func() {
+		close(s.closed)
+		err = errors.Join(s.queue.Close(), s.provider.Close())
+		<-s.done
+	})
+	return err
+}
+
+// toBytes serializes an operation and multihash into a byte slice for storage.
+func toBytes(op byte, key mh.Multihash) []byte {
+	return append([]byte{op}, key...)
+}
+
+// fromBytes deserializes a byte slice back into an operation and multihash.
+func fromBytes(data []byte) (byte, mh.Multihash, error) {
+	op := data[0]
+	h, err := mh.Cast(data[1:])
+	return op, h, err
+}
+
+// getOperations processes a batch of dequeued operations and groups them by
+// type.
+//
+// It discards multihashes from the `StopProviding` operation if
+// `StartProviding` was called after `StopProviding` for the same multihash.
+func getOperations(dequeued [][]byte) ([][]mh.Multihash, error) {
+	ops := [lastOp][]mh.Multihash{}
+	stopProv := make(map[string]struct{})
+
+	for _, bs := range dequeued {
+		op, h, err := fromBytes(bs)
+		if err != nil {
+			return nil, err
+		}
+		switch op {
+		case provideOnceOp:
+			ops[provideOnceOp] = append(ops[provideOnceOp], h)
+		case startProvidingOp, forceStartProvidingOp:
+			delete(stopProv, string(h))
+			ops[op] = append(ops[op], h)
+		case stopProvidingOp:
+			stopProv[string(h)] = struct{}{}
+		}
+	}
+	for hstr := range stopProv {
+		ops[stopProvidingOp] = append(ops[stopProvidingOp], mh.Multihash(hstr))
+	}
+	return ops[:], nil
+}
+
+// worker processes operations from the queue in batches.
+// It runs in a separate goroutine and continues until the provider is closed.
+func (s *SweepingProvider) worker() {
+	defer close(s.done)
+	for {
```

Comment:
Bad if the queue is empty.

Ideally, the queue would send a signal when it receives a first item after being empty.

Reply:
Addressed it in https://github.com/libp2p/go-libp2p-kad-dht/pull/1149/commits/339b0f0b8ddf00bc4e4b3f8c5227486b60a5ab41.